Review 'app.js' proposal for discussion

imalberto opened the pull request:
This PR proposes some ideas to have `app.js` closely resemble what a regular `express` app should look like, and the changes required in mojito to support that paradigm shift.

CHANGELOG:
- Updated all built-in middleware to lazy init
- Moved dispatcher into its own middleware module `middleware/mojito-handler-dispatcher.js`
- Moved stateless functionality into their own module (not related to core mojito): `middleware.js` and `logger.js`
- Updated the `example/newsboxes` and `examples/quickstartguide` examples with their own `app.js` to showcase the use case. These apps can be started with `node app`.
- Fixed middleware related unit tests
- Disabled lib/mojito.js unit tests (for now)

UNIT TESTS RESULTS:

```
********************************************
Consolidated Test Report Summary
********************************************
Total Number of Executed Tests  :633
Total Number of Passed Tests  : 621
Total Number of Failed Tests  : 0
Total Number of Skipped Tests  : 12
********************************************

Total Time of Execution :69.46 seconds
```

imalberto committed:
First pass at prototype for 0.6

imalberto committed:
First pass at prototype for 0.6

imalberto committed:
More changes to lib/mojito.js; Updated newsboxes app

imalberto committed:
More changes to lib/mojito.js; Updated newsboxes app

caridy commented on examples/newsboxes/app.js:
```diff
@@ -0,0 +1,37 @@
+// app.use(require('./middleware/foo.js'));
+
+// Registers Mojito's default
+app.use(app.mojito.registerMiddleware());
```
`registerMiddleware` sounds awkward. should be `registeredMiddleware()` or just `app.mojito.middlewares()`. Now, middlewares should be stateless, which means they should not be bound to `app` at all. This makes me think that probably we should do:

```
app.use(mojito.middlewares());
```

which returns a middleware that groups a set of middleware that are bundled with mojito and takes in consideration any custom middleware in `app.json`. This is exactly what I do in `modown-yui` with `yui.expose()`.

caridy commented on examples/newsboxes/app.js:
```diff
@@ -0,0 +1,37 @@
+/*
+// By doing this, give users more control over which order middleware
+// are registered.
+app.mojito.middleware().forEach(function (mid) {
```
again, this makes a lot of sense, by it will be `mojito.middlewares()`, which is stateless.

ekashida replied:
I also like `mojito.middleware()`. Just a note: the plural of middleware is middleware :)

caridy replied:
lol, thanks @ekashida. I knew something was not right with that "s". jajajaja

caridy commented on examples/newsboxes/app.js:
```diff
@@ -0,0 +1,37 @@
+// are registered.
+app.mojito.middleware().forEach(function (mid) {
+    debug('app.use(): ' + mid);
+    app.use(app.mojito[mid]);
```
probably the default middlewares should be attached to the function itself, something like mojito.middlewares[mid], where the function also acts like an object with members.

caridy commented on examples/newsboxes/app.js:
```diff
@@ -0,0 +1,37 @@
+});
+*/
+
+// In addition to mojito `routes.json`, user can hook up additional 
```
where are the routes from routes.json mounted? I don't see the call here. How can I remove that and do my own thing?

caridy commented on lib/mojito.js:
```diff
@@ -36,55 +50,36 @@ var express = require('express'), // TODO: [Issue 80] go back to connect?
 global._mojito = {};
```
is this really needed?

caridy commented on lib/mojito.js:
```diff
@@ -96,278 +91,236 @@ MojitoServer.MOJITO_MIDDLEWARE = [
- * @type {number}
- */
-MojitoServer.prototype._startupTime = null;
+NOTE: naming is based on ExpressYUIExtension in modown-yui
```
lol

imalberto commented:
Thanks guys for all the feedback. I'll be making changes to this PR based on the ongoing discussion for the next review.

imalberto committed:
- Rewired all mojito- middleware to lazy init
- Moved dispatcher into its own module
- Updated lib/mojito.js with feedback from PR review

imalberto committed:
- Rewired all mojito- middleware to lazy init
- Moved dispatcher into its own module
- Updated lib/mojito.js with feedback from PR review

imalberto committed:
Moved all middleware specific functions to `lib/middleware.js`, which will be mixed in with `lib/mojito.js`

imalberto committed:
Moved all middleware specific functions to `lib/middleware.js`, which will be mixed in with `lib/mojito.js`

imalberto committed:
Added lib/logger as a submodule to lib/mojito

imalberto committed:
Added lib/logger as a submodule to lib/mojito

imalberto committed:
Added app.js for the quickstartguide example;

imalberto committed:
Added app.js for the quickstartguide example;

imalberto committed:
Updated module description

imalberto committed:
Updated module description

imalberto committed:
Removed unused code

caridy commented on examples/newsboxes/app.js:
```diff
@@ -0,0 +1,40 @@
+
+'use strict';
+
+var debug = require('debug')('app:yaml'),
```
should be 'app' instead of 'app:yaml' right?

imalberto replied:
Right.

caridy commented on examples/newsboxes/app.js:
```diff
@@ -0,0 +1,40 @@
+// app.use(require('./middleware/foo.js'));
+
+// Registers Mojito's default
+app.use(mojito.middleware(app));
```
why do we need to pass `app` here? in theory once the first request comes in. you can get a handle on the app by doing `req.app`.

imalberto replied:
`app` is needed to do the `app.use()` to register the middleware, and since `mojito` in this case is stateless, there is no reference to `app`.

caridy commented on examples/newsboxes/app.js:
```diff
@@ -0,0 +1,40 @@
+// are registered.
+// But the recommended way to register all default middleware is:
+// `app.use(mojito.middleware());`
+mojito.defaultMiddleware().forEach(function (mid) {
```
`default` sounds ambiguous here because it is not about default, but those that are set up thru `application.json` isn't it?

imalberto replied:
In this case, the default is really the "mojito built in" middleware.

Question: Do we want to support the "middleware" configuration in application.json in addition to `app.use()` ?

caridy commented on lib/app/middleware/mojito-contextualizer.js:
```diff
@@ -1,15 +1,15 @@
 
-var qs = require('querystring'),
-    logger,
+var debug = require('debug')('middleware:contextualizer'),
```
for debug we should probably use `mojito:middleware:contextualizer`, so you can do any of these:

```
$ export DEBUG=*
$ export DEBUG=mojito:*
$ export DEBUG=mojito:middleware:*
$ export DEBUG=mojito:middleware:contextualizer
```

specially because we will have a lot of components

caridy commented on lib/app/middleware/mojito-contextualizer.js:
```diff
@@ -1,15 +1,15 @@
-var qs = require('querystring'),
-    logger,
+var debug = require('debug')('middleware:contextualizer'),
+    qs = require('querystring'),
```
I heard that `qs` is faster than `querystring`, lets add a card to review this.

caridy commented on lib/app/middleware/mojito-contextualizer.js:
```diff
@@ -44,14 +44,18 @@ function RequestContextualizer() {
 
-        return function(req, res, next) {
+            if (!defaultLang && req.app && req.app.mojito) {
+                defaultLang = (req.app.get('mojito.context') &&
```
hmm, I usually do `app.set('mojito.context')`, because `get` is overloaded, and a mistake can cost you hours of debugging.

caridy commented on lib/app/middleware/mojito-contextualizer.js:
```diff
@@ -44,14 +44,18 @@ function RequestContextualizer() {
-        return function(req, res, next) {
+            if (!defaultLang && req.app && req.app.mojito) {
+                defaultLang = (req.app.get('mojito.context') &&
+                    req.app.get('mojito.context').lang) || DEFAULT_LANG;
```
Is `mojito.context` attached once during the mojito object creation? then it is ok. I will say lets use `mojito context` instead of the dot. More a la express!.

Also, this code can be simplified, not need to ask fro req.app.mojito, and you can use a local variable probably for better readability.

Final question on this: how does that app level setting plays with the request level? do we want that?

caridy commented on lib/app/middleware/mojito-contextualizer.js:
```diff
@@ -156,7 +158,7 @@ RequestContextualizer.prototype = {
-    var contextualizer = new RequestContextualizer();
-    return contextualizer.handle(config.logger, config.context.lang);
+exports = module.exports = function (config) {
+    var m = new RequestContextualizer();
```
why is this still needed? isn't easier to just export the function wrapper?

imalberto replied:
Like `module.exports = new RequestContextualizer().handle();` ?

caridy commented on lib/app/middleware/mojito-handler-dispatcher.js:
```diff
@@ -0,0 +1,71 @@
+
+'use strict';
+
+var debug = require('debug')('middleware:dispatcher'),
```
mojito:middleware:dispatcher

caridy commented on lib/app/middleware/mojito-handler-dispatcher.js:
```diff
@@ -0,0 +1,71 @@
+        
+        if (!store && req.app && req.app.mojito) {
+            store = req.app.get('mojito.store');
+            Y = req.app.get('mojito.Y');
```
eventually this will become `req.app.yui.use('mojito-dispatcher')` as a one time operation. Let's add a big TODO for that.

caridy commented on lib/app/middleware/mojito-handler-dispatcher.js:
```diff
@@ -0,0 +1,71 @@
+
+    var appConfig,
+        store,
+        Y;
```
eventually, Y should not be in this middleware, only a handle to the dispatcher itself.

caridy commented on lib/app/middleware/mojito-handler-static.js:
```diff
@@ -21,14 +21,19 @@
  * Module dependencies.
  */
-var liburl   = require('url'),
+var debug = require('debug')('middleware:static'),
```
mojito:*

caridy commented on lib/app/middleware/mojito-handler-static.js:
```diff
@@ -21,14 +21,19 @@
 
+/**
+
+@module mojito-handler-static
```
I will not review this one deeply because eventually modown-yui should kick in.

caridy commented on lib/app/middleware/mojito-handler-static.js:
```diff
@@ -196,11 +203,25 @@ function staticProvider(store, logger, Y) {
         }
 
+        if (!mojito && req.app && req.app.mojito) {
+            store = req.app.get('mojito.store');
```
I'm worry about using app.set() a lot, specially for things that are very mojito specific. The problem is that we don't know how express is handling that, and how efficient it is. In my opinion those things should be just `req.app.mojito.store` for example.

imalberto replied:
I originally had it in your suggested format. @drewfish and I had a quick discuss about it and thought `app.get()` is more `express` like. But your points are valid. I'll revert it to `req.app.xxx` for now and can revisit later if necessary.

caridy commented on lib/logger.js:
```diff
@@ -0,0 +1,92 @@
+@public
+@param {YUI} Y shared YUI instance on the server
+**/
+module.exports = {
```
good call. the instance created for store as a sandbox should be passed thru this as well, the same for the commands that are creating Y instances.

imalberto replied:
Would need to match the log level from `application.json` - reading the log level without resource store seems like a hack. For now, it appears that resource store always log to "info" level.

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,139 @@
+server instance.
+@type Array
+**/
+Mojito.MOJITO_MIDDLEWARE = [
```
not need to overload `Mojito`, it is only used to hold MOJITO_MIDDLEWARE, remove it.

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,139 @@
+    **/
+    middleware: function (app) {
+
+        module.exports.defaultMiddleware().forEach(function (mid) {
```
not need to walk then and attach them one by one. express supports this:

```
app.use([foo, bar, baz]);
```

so you can return an array of middleware!

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,139 @@
+    @public
+    @return {Array} list of middlewares for a default Mojito application
+    **/
+    defaultMiddleware: function () {
```
this becomes irrelevant if the previous one returns an array.

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,139 @@
+    @param {Object} midConfig configuration object to pass to middleware
+    @param {Array} middleware middleware names
+    **/
+    exposeMiddleware: function () {
```
not so sure what this does!

imalberto replied:
This attaches the middleware to the `mojito` function so that from `app.js`, user can do things like :

`app.use(mojito['mojito-handler-static]);`

caridy commented on lib/mojito.js:
```diff
@@ -1,269 +1,238 @@
- */
-// TODO: Merge what we put on this object with the 'mojito' module/namespace.
-global._mojito = {};
+    app.use(mojito.middleware(app));
```
remove app from here.

caridy commented on lib/mojito.js:
```diff
@@ -1,269 +1,238 @@
-//  ----------------------------------------------------------------------------
+    app = express(); // Mojito instance will be created during that call
 
+    app.use(require('./middleware/my-middleware'));
```
let's not promote this too much. simple simple simple

imalberto replied:
How are we recommending users who have app specific middleware to add them to `express` ?

caridy commented on lib/mojito.js:
```diff
@@ -1,269 +1,238 @@
+//  Util
+//  ----------------------------------------------------------------------------
+// If more helpers are needed, they could potentially be moved to their own
+// `util.js` module later.
```
I already have it in another PR, once I merge it, we can move these methods to that new script.

caridy commented on lib/mojito.js:
```diff
@@ -1,269 +1,238 @@
+The date/time the Mojito object was initialized.
+@type Date
+**/
+Mojito.MOJITO_INIT = new Date().getTime();
```
let's get rid of all those non-sense timers, debug does that for us.

caridy commented on lib/mojito.js:
```diff
@@ -1,269 +1,238 @@
 
-    return middleware;
+    // The application root directory
+    root = process.cwd();
```
watch out with this, manhattan has the wrong cwd.

caridy commented on lib/mojito.js:
```diff
@@ -1,269 +1,238 @@
+
+    options = {};
+    options.port = appConfig.appPort || process.env.PORT || 8666;
+    options.context = { }; // TODO: use correct static context here
```
lets make sure we force runtime: server somewhere

caridy commented on lib/mojito.js:
```diff
@@ -1,269 +1,238 @@
+    options = {};
+    options.port = appConfig.appPort || process.env.PORT || 8666;
+    options.context = { }; // TODO: use correct static context here
+    options.perf =  'perf-log.log'; // TODO: should come from application.json
```
options.perf is going away.

caridy commented on lib/mojito.js:
```diff
@@ -305,411 +274,55 @@ MojitoServer.prototype._configureAppInstance = function(app, options) {
-    // attach middleware pieces
-    this._useMiddleware(app, dispatcher, options.dir, midConfig, middleware);
+    // stash for runtime
+    app.set('mojito.store', store);
```
as I mentioned before, this might be more suitable to be attached to `app.mojito` itself.

caridy commented on lib/mojito.js:
```diff
@@ -305,411 +274,55 @@ MojitoServer.prototype._configureAppInstance = function(app, options) {
+    if (!this.mojito) {
+        // These two probably can be merged.
+        this.mojito = new ExpressMojitoExtension(this);
+        this.mojito._createMojito();
```
why two separate call? what is _createMojito doing that the constructor of the extension can't do?

imalberto committed:
Changes based on most current PR review.

https://github.com/yahoo/mojito/pull/1101

imalberto committed:
Changes based on most current PR review.

https://github.com/yahoo/mojito/pull/1101

imalberto committed:
- update api docs
- linted
- updated middleware initialization to be at runtime instead based on feedback from PR

imalberto committed:
- update api docs
- linted
- updated middleware initialization to be at runtime instead based on feedback from PR

imalberto committed:
Updated api doc.

imalberto committed:
Updated api doc.

imalberto committed:
- Using app.mojito.store (as opposed to app.get('mojito.store))
- Simplify the exported function for lib/mojito.js

imalberto committed:
- Using app.mojito.store (as opposed to app.get('mojito.store))
- Simplify the exported function for lib/mojito.js

imalberto committed:
- linted
- merged the constructor and `createMojito` method in a single call
- updated `app.js` examples to use the revised middleware() api

imalberto committed:
- linted
- merged the constructor and `createMojito` method in a single call
- updated `app.js` examples to use the revised middleware() api

caridy commented on lib/app/middleware/mojito-handler-dispatcher.js:
```diff
@@ -0,0 +1,79 @@
+            log: Y.log
+        });
+
+        outputHandler.page.staticAppConfig = Y.mojito.util.copy(appConfig);
```
In `develop` branch we remove the copy, and instead, we consider `staticAppConfig` an internal object that we will NEVER change.
In fact we made few other changes that will change this block, but we can solve those conflicts when trying to merge this.

imalberto replied:
I'll add a TODO in the source file with your explanation, and resolve it during the next merge.

caridy commented on lib/app/middleware/mojito-handler-static.js:
```diff
@@ -196,11 +205,26 @@ function staticProvider(store, logger, Y) {
             return;
         }
 
+        if (!mojito && req.app && req.app.mojito) {
```
where is `mojito` being set? I don't see it in this file.

imalberto replied:
Thanks! its supposed to be `if (!store && req.app && req.app.mojito) {`

caridy commented on lib/app/middleware/mojito-handler-tunnel-parser.js:
```diff
@@ -17,27 +22,36 @@ var RE_TRAILING_SLASHES = /\/+$/;
+            }
+
+            staticPrefix = staticPrefix || '/static';
+            tunnelPrefix = tunnelPrefix || '/tunnel';
```
I made some changes here as well, merge will have to be done manually~

imalberto replied:
Noted as a comment in the source file to resolve manually after the next merge with `develop`.

caridy commented on lib/app/middleware/mojito-router.js:
```diff
@@ -57,6 +69,16 @@ Router.prototype = {
+                RouteMakerClass = Y.mojito.RouteMaker;
+            }
+
+            routes = store.getRoutes(context);
```
We don't want routes to be contexualized, instead you can compute `routeMaker` once as part of the previous if and use the static context for that.

imalberto replied:
Will update.

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,174 @@
+
+            if (appConfig.middleware && appConfig.middleware.length > 0) {
+                debug('loading app specific middleware first');
+                appDir = process.cwd();
```
this is problematic, the path should come from app.js

imalberto replied:
Will add a note to that effect that the appRootDir should either come from `app.js` or `locator` via an api at some point when `locator` is integrated. This is similar to : https://github.com/yahoo/mojito/pull/1101/files#L18R123

Until such time, `appDir = app.mojito.options.root;`

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,174 @@
+
+            mm.forEach(function (m) {
+                if (MOJITO_MIDDLEWARE.indexOf(m) > -1) {
+                    app.use(middleware[m]);
```
I'm not so sure we should do this. Adding more middleware on demand will cause the actual middleware will not honor the order. Instead, I think we will have to emulate express here, doing our own simple walk on those middleware, and controlling the error and the next. I did that in modown-yui->middleware for the static, please, double check.

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,174 @@
+                        mid  = libpath.join(appDir, m);
+                        debug('app.use() => %s', mid);
+                        mid = require(mid);
+                        app.use(mid());
```
same as above!

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,174 @@
+                        debug('app.use() => %s', mid);
+                        mid = require(mid);
+                        app.use(mid());
+                    } catch (e) {
```
I'm not so sure we need to be defensive here.

imalberto replied:
Will remove the `try/catch` and allow the error bubble up to the application. Better for debugging.

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,174 @@
+            midFactory = require(midName);
+        } catch (e1) {
+            try {
+                midPath = libpath.join(__dirname, 'app', 'middleware', midName);
```
are these really relative to __dirname?

imalberto replied:
Based on the current directory structure where the built-in middleware lives, Yes.

caridy commented on lib/middleware.js:
```diff
@@ -0,0 +1,174 @@
+                midFactory = require(midPath);
+            } catch (e2) {
+                midFactory = null;
+                console.error('failed to attach middleware: ' + midName);
```
instead of console.error we should `throw`

imalberto committed:
More changes based on PR feedback: https://github.com/yahoo/mojito/pull/1101

imalberto committed:
More changes based on PR feedback: https://github.com/yahoo/mojito/pull/1101

imalberto committed:
Updated middleware execution at run time based on PR#1101 feedback.

imalberto committed:
Updated middleware execution at run time based on PR#1101 feedback.

imalberto committed:
Added doc about how objects are exposed via the `app.mojito` instance

imalberto committed:
Added doc about how objects are exposed via the `app.mojito` instance

imalberto committed:
Merge pull request #1123 from imalberto/next

Merging #develop into #next

imalberto committed:
Merged conflicts from origin/fix-appjs into local/fix-appjs

imalberto committed:
- Fixed middleware related tests
- Disabled lib/mojito.js unit tests for now as they have to be rewritten completely.

imalberto committed:
Removed debug statements

caridy commented on lib/mojito.js:
```diff
@@ -305,416 +233,51 @@ MojitoServer.prototype._configureAppInstance = function(app, options) {
-            log: Y.log
-        });
-
-        // storing the static app config as well as contextualized app config per request
```
all these are needed in the new dispatcher middleware.

imalberto replied:
added the missing static and contextualized app config as part of this commit 23e63ce . thanks for catching this.

imalberto committed:
Updated dispatcher after merge with #next as part of PR#1101

caridy commented on lib/mojito.js:
```diff
@@ -1,281 +1,208 @@
- * @return {MojitoServer}
- */
-function MojitoServer(options) {
+Copied from modown-yui.utils
```
This can now be moved into `lib/util.js`

imalberto replied:
updated (commit     c3703c1 )

imalberto committed:
Moved the extend utility fn to lib/utils.js

caridy commented:
I think there is nothing missing at this point.

+1

imalberto committed:
Removed exports assignment after review
